### PR TITLE
CCE-108 Removed address param from CCE API call, retrieved from E&E

### PR DESCRIPTION
**Pull request by CameronPH**
JIRA: https://vasdvp.atlassian.net/browse/CCE-108

Removed the need for address parameters when querying a patient's information. 

For the `smoke-tests` and `regression-tests`, I was unable get the docker test to pass the `Token validation` tests. Derek Brown suggested that a visual verification would be sufficient in this case.

### CODE REVIEW

**Comment by jblefkove-va** [comment]
The `README` should also be updated

File: community-care-eligibility/src/test/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityTest.java [hunk 1]
```diff
@@ -76,18 +82,45 @@ public void audiology() {
...
+                .summary(
+                    EeSummary.builder()
+                        .demographics(
+                            DemographicInfo.builder()
+                                .contactInfo(
+                                    ContactInfo.builder()
+                                        .addresses(
+                                            AddressCollection.builder()
+                                                .address(
+                                                    new ArrayList<AddressInfo>(
```
Comment (jblefkove-va) [inline community-care-eligibility/src/test/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityTest.java]: This `new ArrayList` wrapper is unnecessary, `Arrays.asList` already returns a list.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 2]
```diff
@@ -161,20 +190,16 @@ public CommunityCareEligibilityResponse search(
...
             .build();
-
     if (CollectionUtils.containsAny(codeString, Arrays.asList("G", "N", "H", "X"))) {
-
       return communityCareEligibilityResponse
           .eligible(!codeString.contains("X"))
           .grandfathered(codeString.contains("G"))
           .noFullServiceVaMedicalFacility(codeString.contains("N"));
     }
-
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: The order of fields in the response should roughhly reflect the order in which the algorithm uses them. Now that address is no longer part of the request, that suggests a couple things:

* `patientAddress` should move out of `patientRequest` to the top level of the response

* `patientAddress` should appear in the response after `noFullServiceVaMedicalFacility`, because we don't need to look at the address in cases where the veteran has an E&E code that already makes them explicitly eligible or ineligible. 

* The computation of the address should move down to this line.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 3]
```diff
@@ -140,14 +176,7 @@ public CommunityCareEligibilityResponse search(
...
-    Address patientAddress =
-        Address.builder()
-            .street(street.trim())
-            .city(city.trim())
-            .state(state.toUpperCase(Locale.US).trim())
-            .zip(zip.trim())
-            .build();
+    Address patientAddress = parsePatientAddress(patientResidentialAddressInfo(response));
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: `parsePatientAddress` throws a null pointer exception if no address is found in E&E. We should handle that more gracefully. Let's add a new exception type to throw if no residential address is found. We should be similarly restrictive for the case where the E&E address is available, but incomplete. (Missing any of the four fields.)

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 4]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+              .append("-")
+              .append(patientResidentialAddressInfo.getZipPlus4())
+              .toString());
+    }
+    return patientAddress;
+  }
+
+  private AddressInfo patientResidentialAddressInfo(GetEESummaryResponse response) {
+    for (AddressInfo info :
+        response.getSummary().getDemographics().getContactInfo().getAddresses().getAddress()) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: In general, we need to be cautious when dealing with data from other APIs. There are no guarantees about availability, so any of the calls in this long chain could potentially be null, and we should handle that. Take a look at the other method in this class that extracts the eligibility codes from the eeResponse.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 5]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+            .state(patientResidentialAddressInfo.getState())
+            .zip(
+                (new StringBuilder()).append(patientResidentialAddressInfo.getZipCode()).toString())
+            .build();
+    StringBuilder streetBuilder = new StringBuilder();
+    for (String line :
+        Arrays.asList(
+            patientResidentialAddressInfo.getLine1(),
+            patientResidentialAddressInfo.getLine2(),
+            patientResidentialAddressInfo.getLine3())) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: This could be done more elegantly with a stream using `Collectors.join(", ")`. (i.e. create a stream with the three lines, trim  them, discard the blanks, then concatenate)

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 6]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
         .collect(Collectors.toList());
   }
 
+  private Address parsePatientAddress(AddressInfo patientResidentialAddressInfo) {
+    Address patientAddress =
+        Address.builder()
+            .city(patientResidentialAddressInfo.getCity())
+            .state(patientResidentialAddressInfo.getState())
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: State should be shifted to uppercase.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 7]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
         .collect(Collectors.toList());
   }
 
+  private Address parsePatientAddress(AddressInfo patientResidentialAddressInfo) {
+    Address patientAddress =
+        Address.builder()
+            .city(patientResidentialAddressInfo.getCity())
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: As a good practice, these strings from the other API should be trimmed. Take a look at some of the methods in the FacilityTransformer class.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 8]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+            .state(patientResidentialAddressInfo.getState())
+            .zip(
+                (new StringBuilder()).append(patientResidentialAddressInfo.getZipCode()).toString())
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: You don't need a stringbuilder to only append one thing to it.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 9]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+            patientResidentialAddressInfo.getLine1(),
+            patientResidentialAddressInfo.getLine2(),
+            patientResidentialAddressInfo.getLine3())) {
+      if ((streetBuilder.length() != 0) && !line.isEmpty()) {
+        streetBuilder.append(" ");
+      }
+      streetBuilder.append(line);
+    }
+    patientAddress.street(streetBuilder.toString());
+    if (patientResidentialAddressInfo.getZipPlus4() != null) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: All of the logic for computing the zip code, including the extension, should be together.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 10]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+          (new StringBuilder())
+              .append(patientResidentialAddressInfo.getZipCode())
+              .append("-")
+              .append(patientResidentialAddressInfo.getZipPlus4())
+              .toString());
+    }
+    return patientAddress;
+  }
+
+  private AddressInfo patientResidentialAddressInfo(GetEESummaryResponse response) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: This method and the one above can be static.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 11]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
         .collect(Collectors.toList());
   }
 
+  private Address parsePatientAddress(AddressInfo patientResidentialAddressInfo) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: The parameter name is unnecessarily long and makes things a little harder to read. In the context of this method, we're just converting some EE address to our model of address. It isn't specially for a patient, or residential.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 12]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
...
+      }
+      streetBuilder.append(line);
+    }
+    patientAddress.street(streetBuilder.toString());
+    if (patientResidentialAddressInfo.getZipPlus4() != null) {
+      patientAddress.zip(
+          (new StringBuilder())
+              .append(patientResidentialAddressInfo.getZipCode())
+              .append("-")
+              .append(patientResidentialAddressInfo.getZipPlus4())
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: StringBuilder is mainly useful for building strings in loops, or for situations where you're building a string with complicated logic with a lot of branches. It's not really necessary here.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 13]
```diff
@@ -104,25 +104,61 @@ public CommunityCareEligibilityV0ApiController(
         .collect(Collectors.toList());
   }
 
+  private Address parsePatientAddress(AddressInfo patientResidentialAddressInfo) {
+    Address patientAddress =
+        Address.builder()
+            .city(patientResidentialAddressInfo.getCity())
+            .state(patientResidentialAddressInfo.getState())
+            .zip(
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: If `zipCode` is blank, we should use `postalCode` instead. And if that is blank, fall back to the deprecated `zipcode`.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java (hunk 5 above)
Comment (CameronPH) [reply community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: After reviewing how the street is concatenated from three different lines in `FacilityTransformer.java`, I can just use a combination of `StringUtils.trimToEmpty()` with a regular string concatenation operator.

The reason it was set up this way before was I did not know how to prevent attempting to access one of the streets if it was `null`. But `StringUtils.trimToEmpty()` converts `null` strings to the empty string.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java (hunk 4 above)
Comment (CameronPH) [reply community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: In regards to the cases where the method doesn't return an `AddressInfo` object with `addressTypeCode = "Residential"`. What should this method return? I could return a new instantiation of an `AddressInfo` object, and all fields that are set to `null` will be converted to the empty string through `StringUtils.trimToEmpty()`. So it wouldn't throw any errors and we would just get an `Address` object with all empty strings.

Is that acceptable or do you want an error message/code of some kind to be returned when no "Residential" `addressTypeCode` is found.

Comment (CameronPH) [reply community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: Disregard my statement, you've already stated that you want an new exception type for that exact case.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 14]
```diff
@@ -66,6 +67,64 @@ public CommunityCareEligibilityV0ApiController(
...
+                StringUtils.trim(
+                    StringUtils.trimToEmpty(addressInfo.getLine1())
+                        + " "
+                        + StringUtils.trimToEmpty(addressInfo.getLine2())
+                        + " "
+                        + StringUtils.trimToEmpty(addressInfo.getLine3())))
+            .build();
+    if (addressInfo.getZipCode().isEmpty()) {
+      if (addressInfo.getPostalCode().isEmpty()) {
+        patientAddress.zip(StringUtils.trimToEmpty(addressInfo.getZipCode()));
```
Comment (CameronPH) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: You stated that if neither the zipCode nor postalCode we filled in, to fall back on the zipCode. Would it not be better throw an MissingAddressInformationException() instead?

File: README.md [hunk 15]
```diff
@@ -56,7 +56,7 @@ facilities that satisfy the access standards.
 Sample request:
 
 ```
-https://foo.com/community-care/v0/eligibility/search?patient=011235813V213455&street=742%20Evergreen%20Terrace&city=Springfield&state=KY&zip=89144&serviceType=primarycare
+https://foo.com/community-care/v0/eligibility/search?patient=011235813V213455&serviceType=primarycare
```
Comment (jblefkove-va) [inline README.md]: There is a line above that mentions requesting patient home address that should also be removed. And the sample response below should be updated.

File: community-care-eligibility-api/src/main/java/gov/va/api/health/communitycareeligibility/api/CommunityCareEligibilityResponse.java [hunk 16]
```diff
@@ -23,6 +23,8 @@
 
   PatientRequest patientRequest;
 
+  Address patientAddress;
```
Comment (jblefkove-va) [inline community-care-eligibility-api/src/main/java/gov/va/api/health/communitycareeligibility/api/CommunityCareEligibilityResponse.java]: The order of fields here should match the order that the algorithm uses them... so put address after `noFullServiceVaMedicalFacility`

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java [hunk 17]
```diff
@@ -27,4 +27,16 @@
       super("Unknown service type: " + serviceType);
     }
   }
+
+  static final class MissingResidentialAddressException extends NullPointerException {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java]: Keep the exceptions alphabetized. And these should extend `RuntimeException`

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java [hunk 18]
```diff
@@ -27,4 +27,16 @@
...
+    MissingResidentialAddressException() {
+      super("No residential address found");
+    }
+  }
+
+  static final class MissingAddressInformationException extends NullPointerException {
+    MissingAddressInformationException() {
+      super("Residential address has incomplete information");
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java]: To help with debugging, the incomplete address should be included on the exception. For example, if only the zip code is missing, showing the address object with street/city/state populated is helpful. So I would move the check for missing data to the end of the `parsePatientAddress` method, and throw the `MissingAddressInformationException` at that time

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java [hunk 19]
```diff
@@ -27,4 +27,16 @@
       super("Unknown service type: " + serviceType);
     }
   }
+
+  static final class MissingResidentialAddressException extends NullPointerException {
+    MissingResidentialAddressException() {
+      super("No residential address found");
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java]: Include the patient ICN in this error message.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java [hunk 20]
```diff
@@ -27,4 +27,16 @@
...
+    }
+  }
+
+  static final class MissingAddressInformationException extends NullPointerException {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java]: At a glance `MissingResidentialAddress` and `MissingAddressInformation` are easy to get mixed up. How about calling this `IncompleteAddressException`.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 21]
```diff
@@ -65,13 +67,70 @@ public CommunityCareEligibilityV0ApiController(
...
   }
 
+  @SneakyThrows
+  private static Address parsePatientAddress(AddressInfo addressInfo) {
+    if (addressInfo == null) {
+      throw new Exceptions.MissingResidentialAddressException();
+    }
+    Address patientAddress =
+        Address.builder()
+            .city(StringUtils.trim(addressInfo.getCity()))
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: Use static imports to make all these calls to `StringUtils` methods more readable.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 22]
```diff
@@ -65,13 +67,67 @@ public CommunityCareEligibilityV0ApiController(
...
+        || response.getSummary() == null
+        || response.getSummary().getDemographics() == null
+        || response.getSummary().getDemographics().getContactInfo() == null
+        || response.getSummary().getDemographics().getContactInfo().getAddresses() == null) {
+      return null;
+    }
+    for (AddressInfo info :
+        response.getSummary().getDemographics().getContactInfo().getAddresses().getAddress()) {
+      if ("Residential".equals(info.getAddressTypeCode())) {
+        return info;
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: This pattern of finding the first matching item in a list can be done more elegantly with a stream. Take a look at https://github.com/department-of-veterans-affairs/health-apis-data-query/blob/master/data-query-tests/src/main/java/gov/va/api/health/dataquery/tests/crawler/ResourceDiscovery.java#L112

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java [hunk 23]
```diff
@@ -16,6 +17,31 @@
     }
   }
 
+  static final class IncompleteAddressException extends RuntimeException {
+    IncompleteAddressException(Address patientAddress) {
+      super(
+          "Residential address has incomplete information"
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/Exceptions.java]: The extra formatting isn't really necessary. You can do `"Residential address has incomplete information: " + patientAddress` and rely on the address `toString` method.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 24]
```diff
@@ -65,13 +67,67 @@ public CommunityCareEligibilityV0ApiController(
...
+    if (addressInfo.getZipCode().isEmpty()) {
+      if (addressInfo.getPostalCode().isEmpty()) {
+        patientAddress.zip(trimToEmpty(addressInfo.getZipCode()));
+      } else {
+        patientAddress.zip(trimToEmpty(addressInfo.getPostalCode()));
+      }
+    } else {
+      patientAddress.zip(trimToEmpty(addressInfo.getZipCode()));
+    }
+    if (addressInfo.getZipPlus4() != null) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: There should be a check that the extension is only added if the base zip code is not blank. Otherwise `-1234` will look like a valid zip code, but is not.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 25]
```diff
@@ -65,13 +67,67 @@ public CommunityCareEligibilityV0ApiController(
...
+                trimToEmpty(
+                    trimToEmpty(addressInfo.getLine1())
+                        + " "
+                        + trimToEmpty(addressInfo.getLine2())
+                        + " "
+                        + trimToEmpty(addressInfo.getLine3())))
+            .build();
+    if (addressInfo.getZipCode().isEmpty()) {
+      if (addressInfo.getPostalCode().isEmpty()) {
+        patientAddress.zip(trimToEmpty(addressInfo.getZipCode()));
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: If `zipCode` and `postalCode` are both blank, you can fall back on `zipcode` (all lowercase) as a last attempt to populate the zip code.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 26]
```diff
@@ -65,13 +67,67 @@ public CommunityCareEligibilityV0ApiController(
     this.facilitiesClient = facilitiesClient;
   }
 
+  @SneakyThrows
+  private static Address parsePatientAddress(AddressInfo addressInfo) {
+    Address patientAddress =
+        Address.builder()
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: It's confusing that the address is built up front, but not all of the fields are populated. Calculate the zip code first and store that in a local variable, then build the address and set all four fields at the same time.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 27]
```diff
@@ -65,13 +67,67 @@ public CommunityCareEligibilityV0ApiController(
...
+            .state(trimToEmpty(addressInfo.getState().toUpperCase()))
+            .street(
+                trimToEmpty(
+                    trimToEmpty(addressInfo.getLine1())
+                        + " "
+                        + trimToEmpty(addressInfo.getLine2())
+                        + " "
+                        + trimToEmpty(addressInfo.getLine3())))
+            .build();
+    if (addressInfo.getZipCode().isEmpty()) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: These checks are easier to follow if you initialize `zip` to `zipCode` and then override it if `zip` is empty:

```
String zip = trimToEmpty(addressInfo.getZipCode());
if (zip.isEmpty()) {
  zip = trimToEmpty(addressInfo.getPostalCode());
} 
```

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 28]
```diff
@@ -65,13 +68,73 @@ public CommunityCareEligibilityV0ApiController(
...
+      zip = trimToEmpty(addressInfo.getPostalCode());
+    }
+
+    if (addressInfo.getZipPlus4() != null && !zip.isEmpty()) {
+      zip = zip + "-" + trimToEmpty(addressInfo.getZipPlus4());
+    }
+    Address patientAddress =
+        Address.builder()
+            .city(trimToEmpty(addressInfo.getCity()))
+            .state(trimToEmpty(addressInfo.getState().toUpperCase()))
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: This is a `NullPointerException` if `getState()` is null.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 29]
```diff
@@ -65,13 +68,73 @@ public CommunityCareEligibilityV0ApiController(
     this.facilitiesClient = facilitiesClient;
   }
 
+  @SneakyThrows
+  private static Address parsePatientAddress(AddressInfo addressInfo) {
+    String zip = trimToEmpty(addressInfo.getZipCode());
+    if (zip.isEmpty()) {
+      if (addressInfo.getPostalCode().isEmpty()) {
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: This is a `NullPointerException` if `getPostalCode()` is null.

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 30]
```diff
@@ -65,13 +68,73 @@ public CommunityCareEligibilityV0ApiController(
...
+        zip = trimToEmpty(addressInfo.getZipcode());
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: If this line executes to set `zip` to `Zipcode` the next line will still execute and set it to the blank `postalCode`

File: community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java [hunk 31]
```diff
@@ -65,13 +68,73 @@ public CommunityCareEligibilityV0ApiController(
...
+      }
+      zip = trimToEmpty(addressInfo.getPostalCode());
+    }
+
+    if (addressInfo.getZipPlus4() != null && !zip.isEmpty()) {
+      zip = zip + "-" + trimToEmpty(addressInfo.getZipPlus4());
```
Comment (jblefkove-va) [inline community-care-eligibility/src/main/java/gov/va/api/health/communitycareeligibility/service/CommunityCareEligibilityV0ApiController.java]: In the (admittedly unlikely) case that `zipPlus4` is not null, but is blank, the `zip` will be something like `12345-`, which isn't ideal.

**Comment by jblefkove-va** [comment]
To test for some of the `NullPointerException`s I pointed out, I recommend adding a unit test with an address that has the type set to `Residential` and everything else `null`. And verify that the test throws the `IncompleteAddressException`.
